pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp

```diff
@@ -445,6 +446,99 @@ template <class Precision> class StateVectorKokkos {
         }
     };
 
+    /**
+     * @brief Utility method for samples.
+     *
+     * @param num_samples Number of Samples
+     *
+     * @return Kokkos::View<size_t *> to the samples.
+     * Each sample has a length equal to the number of qubits. Each sample can
+     * be accessed using the stride sample_id*num_qubits, where sample_id is a
+     * number between 0 and num_samples-1.
+     */
+
+    auto generate_samples(size_t num_samples)->Kokkos::View<size_t *> {
+
+        const size_t num_qubits = getNumQubits(); // no need for dec
+
+        const size_t N = getLength();
+
+        const size_t Nsqrt =
+            static_cast<size_t>(std::ceil(std::sqrt(static_cast<double>(N))));
+
+        Kokkos::View<Kokkos::complex<Precision> *> arr_data = getData();
+
+        Kokkos::View<Precision *> probabilities("probabilities", N);
+
+        Kokkos::View<Precision *> cld("cld", N);
+
+        Kokkos::View<size_t *> samples("num_samples", num_samples * num_qubits);
+
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, N),
+            KOKKOS_LAMBDA(const size_t &i) {
+                probabilities[i] = imag(arr_data[i]) * imag(arr_data[i]) +
+                                   real(arr_data[i]) * real(arr_data[i]);
+            });
+
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, Nsqrt),
+            KOKKOS_LAMBDA(const size_t &i) {
+                for (size_t j = 0; j < Nsqrt; j++) {
+                    size_t idx = j + i * Nsqrt;
+                    if (j == 0)
+                        cld[idx] = probabilities[idx];
+                    else
+                        cld[idx] = probabilities[idx] + cld[idx - 1];
+                }
+            });
+
+        for (size_t i = 1; i < Nsqrt; i++) {
```

Comment:
What is this loop for?

Reply:
Yeah I'm not sure this is the fastest way to do things, but I could be wrong and a comparison would be nice if you really have faith in it. I would suggest just splitting up num_samples between threads and each thread runs bisection on the cumulative distribution array. So each thread runs O(logN) for a sample and using the parallel_scan function from Kokkos to compute the cum dist array in parallel. Maybe I'm missing something though.

Reply:
Good point! Worth to have a try later!

Reply:
done!